File: frontend/src/api/progressTracking.js
const API_BASE_URL = 'http://localhost:8000';

export const ProgressTrackingService = {
    /**
     * Tạo ghi nhận tiến bộ/hạn chế học tập
     * @param {Object} data - Dữ liệu ghi nhận
     * @param {string} tutorId - ID của tutor
     */
    async createProgressTracking(data, tutorId) {
        try {
            const response = await fetch(`${API_BASE_URL}/progress-tracking/create?tutor_id=${tutorId}`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify(data),
            });

            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.detail || "Tạo ghi nhận tiến bộ thất bại");
            }

            return await response.json();
        } catch (error) {
            console.error("Lỗi khi tạo ghi nhận tiến bộ:", error);
            throw error;
        }
    },

    /**
     * Cập nhật ghi nhận tiến bộ
     * @param {string} trackingId - ID của ghi nhận
     * @param {Object} data - Dữ liệu cập nhật
     * @param {string} tutorId - ID của tutor
     */
    async updateProgressTracking(trackingId, data, tutorId) {
        try {
            const response = await fetch(`${API_BASE_URL}/progress-tracking/${trackingId}?tutor_id=${tutorId}`, {
                method: "PUT",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify(data),
            });

            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.detail || "Cập nhật ghi nhận tiến bộ thất bại");
            }

            return await response.json();
        } catch (error) {
            console.error("Lỗi khi cập nhật ghi nhận tiến bộ:", error);
            throw error;
        }
    },

    /**
     * Xóa ghi nhận tiến bộ
     * @param {string} trackingId - ID của ghi nhận
     * @param {string} tutorId - ID của tutor
     */
    async deleteProgressTracking(trackingId, tutorId) {
        try {
            const response = await fetch(`${API_BASE_URL}/progress-tracking/${trackingId}?tutor_id=${tutorId}`, {
                method: "DELETE",
            });

            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.detail || "Xóa ghi nhận tiến bộ thất bại");
            }

            return await response.json();
        } catch (error) {
            console.error("Lỗi khi xóa ghi nhận tiến bộ:", error);
            throw error;
        }
    },

    /**
     * Lấy tất cả ghi nhận tiến bộ của một mentee
     * @param {string} menteeId - ID của mentee
     * @param {string} tutorId - ID của tutor (tùy chọn)
     */
    async getMenteeProgressTrackings(menteeId, tutorId = null) {
        try {
            const url = tutorId
                ? `${API_BASE_URL}/progress-tracking/mentee/${menteeId}?tutor_id=${tutorId}`
                : `${API_BASE_URL}/progress-tracking/mentee/${menteeId}`;
            const response = await fetch(url, {
                method: "GET",
            });

            if (!response.ok) {
                throw new Error(`Lỗi HTTP: ${response.status}`);
            }

            return await response.json();
        } catch (error) {
            console.error("Lỗi khi lấy ghi nhận tiến bộ của mentee:", error);
            return [];
        }
    },

    /**
     * Lấy tất cả ghi nhận tiến bộ của một session
     * @param {string} sessionId - ID của session
     * @param {string} tutorId - ID của tutor (tùy chọn)
     */
    async getSessionProgressTrackings(sessionId, tutorId = null) {
        try {
            const url = tutorId
                ? `${API_BASE_URL}/progress-tracking/session/${sessionId}?tutor_id=${tutorId}`
                : `${API_BASE_URL}/progress-tracking/session/${sessionId}`;
            const response = await fetch(url, {
                method: "GET",
            });

            if (!response.ok) {
                throw new Error(`Lỗi HTTP: ${response.status}`);
            }

            return await response.json();
        } catch (error) {
            console.error("Lỗi khi lấy ghi nhận tiến bộ của session:", error);
            return [];
        }
    },

    /**
     * Lấy tất cả ghi nhận tiến bộ của một tutor
     * @param {string} tutorId - ID của tutor
     */
    async getTutorProgressTrackings(tutorId) {
        try {
            const response = await fetch(`${API_BASE_URL}/progress-tracking/tutor/${tutorId}`, {
                method: "GET",
            });

            if (!response.ok) {
                throw new Error(`Lỗi HTTP: ${response.status}`);
            }

            return await response.json();
        } catch (error) {
            console.error("Lỗi khi lấy ghi nhận tiến bộ của tutor:", error);
            return [];
        }
    },

    /**
     * Lấy tổng hợp tiến bộ của một mentee
     * @param {string} menteeId - ID của mentee
     * @param {string} tutorId - ID của tutor (tùy chọn)
     */
    async getMenteeProgressSummary(menteeId, tutorId = null) {
        try {
            const url = tutorId
                ? `${API_BASE_URL}/progress-tracking/mentee/${menteeId}/summary?tutor_id=${tutorId}`
                : `${API_BASE_URL}/progress-tracking/mentee/${menteeId}/summary`;
            const response = await fetch(url, {
                method: "GET",
            });

            if (!response.ok) {
                throw new Error(`Lỗi HTTP: ${response.status}`);
            }

            return await response.json();
        } catch (error) {
            console.error("Lỗi khi lấy tổng hợp tiến bộ:", error);
            return null;
        }
    },

    /**
     * Lấy danh sách sessions của tutor kèm phản hồi từ mentee
     * @param {string} tutorId - ID của tutor
     */
    async getTutorSessionsWithFeedback(tutorId) {
        try {
            const response = await fetch(`${API_BASE_URL}/progress-tracking/tutor/${tutorId}/sessions-with-feedback`, {
                method: "GET",
            });

            if (!response.ok) {
                throw new Error(`Lỗi HTTP: ${response.status}`);
            }

            return await response.json();
        } catch (error) {
            console.error("Lỗi khi lấy sessions với phản hồi:", error);
            return [];
        }
    }
};

export default ProgressTrackingService;

File: backend/data/progress_tracking.py
"""
Database giả lập cho Progress Tracking (Theo dõi tiến bộ học tập)
"""
from typing import Dict, List, Optional
import uuid
from datetime import datetime

# Database giả lập: Progress Tracking
fake_progress_tracking_db: Dict[str, Dict] = {}

def get_progress_tracking(tracking_id: str) -> Optional[Dict]:
    """Lấy thông tin theo dõi tiến bộ theo ID"""
    return fake_progress_tracking_db.get(tracking_id)

def get_mentee_progress_trackings(mentee_id: str) -> List[Dict]:
    """Lấy tất cả ghi nhận tiến bộ của một mentee"""
    trackings = []
    for tracking in fake_progress_tracking_db.values():
        if tracking.get("mentee_id") == mentee_id:
            trackings.append(tracking)
    # Sắp xếp theo thời gian (mới nhất trước)
    trackings.sort(key=lambda x: x.get("created_at", ""), reverse=True)
    return trackings

def get_session_progress_trackings(session_id: str) -> List[Dict]:
    """Lấy tất cả ghi nhận tiến bộ của một session"""
    trackings = []
    for tracking in fake_progress_tracking_db.values():
        if tracking.get("session_id") == session_id:
            trackings.append(tracking)
    # Sắp xếp theo thời gian (mới nhất trước)
    trackings.sort(key=lambda x: x.get("created_at", ""), reverse=True)
    return trackings

def get_tutor_progress_trackings(tutor_id: str) -> List[Dict]:
    """Lấy tất cả ghi nhận tiến bộ của một tutor"""
    trackings = []
    for tracking in fake_progress_tracking_db.values():
        if tracking.get("tutor_id") == tutor_id:
            trackings.append(tracking)
    # Sắp xếp theo thời gian (mới nhất trước)
    trackings.sort(key=lambda x: x.get("created_at", ""), reverse=True)
    return trackings

def create_progress_tracking(tracking_data: Dict) -> Dict:
    """Tạo ghi nhận tiến bộ mới"""
    tracking_id = f"PT{str(uuid.uuid4())[:8].upper()}"
    tracking_data["tracking_id"] = tracking_id
    tracking_data["created_at"] = datetime.now().isoformat()
    tracking_data["updated_at"] = None
    fake_progress_tracking_db[tracking_id] = tracking_data
    return tracking_data

def update_progress_tracking(tracking_id: str, update_data: Dict) -> Optional[Dict]:
    """Cập nhật ghi nhận tiến bộ"""
    tracking = fake_progress_tracking_db.get(tracking_id)
    if not tracking:
        return None
    
    for key, value in update_data.items():
        if value is not None:
            tracking[key] = value
    
    tracking["updated_at"] = datetime.now().isoformat()
    return tracking

def delete_progress_tracking(tracking_id: str) -> bool:
    """Xóa ghi nhận tiến bộ"""
    if tracking_id in fake_progress_tracking_db:
        del fake_progress_tracking_db[tracking_id]
        return True
    return False

